### ISSUE TITLE
Using react hooks with ink: Invariant Violation

### ISSUE THREAD

**Issue by shlomokraus** [vadimdemedes/ink#108]
Hey,

I am trying to use the new react hooks with Ink components, but I get the error: 

```
Invariant Violation: Hooks can only be called inside the body of a function component.
```

I am using the hooks inside a function component, so I guess it is related to Ink implementation? 

Here is the code: 

```ts
import React, {useState} from 'react'
import {h, render, Component} from "ink";

function useTest() {
	const [value, update] = useState(Date.now());
	return [value, update]
}

const Demo = () => {
	const [value, update] = useTest();

	setInterval(() => {
		update(Date.now())
	}, 1000)

	return <div>Hello: {value}</div>

}

class Wrapper extends Component<any, any>{
	
	render() {
		return <Demo />
	}
}
render((<Wrapper />));
```

**Comment by mAAdhaTTah** [comment]
The issue is Ink is getting pegged to a version of react-reconciler that doesn't support hooks yet. I was able to get it working by using yarn and adding these resolutions to my package.json:

```json
"resolutions": {
    "react-reconciler": "0.18.0-alpha.2",
    "scheduler": "0.12.0-alpha.3"
}
```

Those alpha versions have the hooks code. Then you can import the hooks and use them. I converted the README:

```jsx
import React, { useState, useEffect } from 'react';
import { render, Color } from 'ink';

const Counter = () => {
  const [i, setI] = useState(0);

  useEffect(() => {
    const timer = setInterval(() => setI(i + 1), 100);
    return () => clearInterval(timer);
  });

  return <Color green>{i} tests passed</Color>;
};

render(<Counter />);
```

**Comment by shlomokraus** [comment]
Doesn't work for me, when I am doing yarn install it says
`warning Resolution field "scheduler@0.12.0-alpha.3" is incompatible with requested version "scheduler@^0.11.0-alpha.0"`
And still getting the invariant violation. 

Can you please share the complete package.json please?

**Comment by shlomokraus** [comment]
Which version of react are you using?

**Comment by mAAdhaTTah** [comment]
It looks like your "dependencies" field has a version of `scheduler`, which it shouldn't. You should just use the "resolutions" field. Also, I tested this prior to the latest hooks version, so I'm not sure what version those should be now.

**Comment by mAAdhaTTah** [comment]
This is what I had when it worked:

```
{
  "name": "test-cli-app",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "ink": "2.0.0-10",
    "react": "16.7.0-alpha.2"
  },
  "devDependencies": {
    "@babel/core": "^7.2.2",
    "@babel/node": "^7.2.2",
    "@babel/plugin-transform-modules-commonjs": "^7.2.0",
    "@babel/preset-react": "^7.0.0",
    "@babel/preset-typescript": "^7.1.0",
    "@types/chalk": "^2.2.0",
    "@types/react": "^16.7.18",
    "prettier": "^1.15.3",
    "typescript": "^3.2.2"
  },
  "resolutions": {
    "react-reconciler": "0.18.0-alpha.2",
    "scheduler": "0.12.0-alpha.3"
  }
}
```

```
{
  "presets": ["@babel/preset-react", "@babel/preset-typescript"],
  "plugins": ["@babel/plugin-transform-modules-commonjs"]
}
```

I know they published a new alpha version, so I don't know if that changes things, but it did work at the time.

**Comment by mAAdhaTTah** [comment]
Oh, wait, you're using ink's `h`, rather than React.createElement. That's your issue.

**Comment by shlomokraus** [comment]
I thought h pragma is required. What is the correct setup then?

**Comment by mAAdhaTTah** [comment]
Ohhh my mistake, here's the issue: the current version of Ink is a custom implementation, so it's not tied to React in particular so much as it's inspired by it. The next version (2.0, as seen on the `next` branch) uses React directly, which is what you'll need to make hooks work. Since hooks are coupled to React, you can't use hooks with the current version of Ink at all.

If you switch to the `next` branch on GitHub, you can see the install & usage instructions. You'll also notice in my pkg.json that I'm using that version of Ink in my hooks-working test project.

**Comment by vadimdemedes** [comment]
Thanks everyone for pitching in on this issue! If I'm not mistaken, release of React Hooks is around the corner and they should release stable versions of their packages, which I will be able to safely upgrade Ink to. Afterwards hooks will be usable with Ink.

**Comment by fnky** [comment]
Though, I have noticed that hooks doesn't quite work the way you're used to on browser/native environments, due to the nature of how the lifecycle/lifetime of a program works in a command-line. 

For example, I tried to port `ink-text-input` using hooks and can't get input to work properly. I have tried with `useEffect` and `process.on('keypress')` as well as `process.stdin.resume()`.

Without using `resume` to actively listen for user input, it will run, but the handler won't be called, even though the listener for the 'keypress' event is registered in `useEffect` like so:

```
useEffect(() => {
    // Using resume will ask for user input and "pause" the program further and won't
    // continue to render until that is done.
    // process.stdin.resume();
    process.stdin.on('keypress', handleKeyPress);

    return () => {
      process.stdin.removeListener('keypress', handleKeyPress);
    };
  }, []);
};
```

The `useEffect` hook is run once every render, but will only run the handler once as I pass an empty array `[]`, so that we only register the listeners once and not every render (e.g. when `value` is updated from the outside).

The program will just exit, as there's nothing that tells it to stop and wait for user input before continuing.

I also tried to use the `StdinContext` with `useContext` in place of `process.stdin`, but that had the same result. Although using `stdinContext.setRawMode` will somehow render the component and seem to listen for user input, as the program keeps running but the handler passed to keypress isn't being executed.

**Comment by fnky** [comment]
So I just investigated the keypress event thing, and I got it sorted out by using `readline.emitKeypressEvents` to emit keypress events for `process.stdin` and it seemed to work.

**Comment by Jessidhia** [comment]
Yeah, older versions of `scheduler` that are used by `useEffect`, when run in not-a-browser, will just set a timeout that expires on infinity. React _will_ force-execute all pending `useEffect`s before doing anything else that is new, but until then it'll just be pending forever.

Current `scheduler` (and the one that will be used by 16.8.0) will always use timeouts of `0` and merely order things according to priority.

**Comment by mAAdhaTTah** [comment]
@Jessidhia I think this is related to Ink not implementing `schedulePassiveEffects` or `cancelPassiveEffects` in its reconciler. With the now-released version of Hooks, I'm getting errors related to that function being missing. Anyone else play with this yet?

**Comment by Jessidhia** [comment]
A very bare implementation would be `setTimeout` / `cancelTimeout`; probably the best if there is no way to know when/if the terminal emulator is done updating the screen.

**Comment by mAAdhaTTah** [comment]
Cool, I'm working on a CLI tool with Ink@next so I'll PR this upstream. Ink also would need to bump some dep versions to get it working with Hooks again.

**Comment by zkat** [comment]
This should be fixed now ✌️ See #124

**Comment by vadimdemedes** [comment]
Hooks are working now thanks to @zkat!
